File: web/src/services/catalogService.ts
import apiClient from './apiClient';
import { Album, Artist, CatalogResults, Track } from '../types/catalog';

const withResults = <T>(response: unknown): T[] => {
  if (!response || typeof response !== 'object') {
    return [];
  }

  const data = response as { results?: T[] };
  return data.results ?? [];
};

const fetchCollection = async <T>(
  resource: 'albums' | 'artists' | 'tracks' | 'genres',
  token: string,
  query: string,
) => {
  const queryParams = query
    ? {
        search: query,
        q: query,
        external: 'true',
      }
    : undefined;
  const response = await apiClient.get(`/api/v1/${resource}/`, {
    token,
    query: queryParams,
  });
  return withResults<T>(response);
};

export const fetchAllResources = async (token: string, query: string): Promise<CatalogResults> => {
  const [albums, artists, tracks] = await Promise.all([
    fetchCollection<Album>('albums', token, query),
    fetchCollection<Artist>('artists', token, query),
    fetchCollection<Track>('tracks', token, query),
  ]);

  return {
    genres: [],
    albums,
    artists,
    tracks,
  };
};

File: mobile/ios/juke-iOS/Services/CatalogService.swift
import Foundation

final class CatalogService {
    private let client: APIClient

    init(client: APIClient = .shared) {
        self.client = client
    }

    func fetchCatalog(token: String, query: String) async throws -> CatalogResults {
        async let artists: [Artist] = fetchCollection(resource: "artists", token: token, query: query)
        async let albums: [Album] = fetchCollection(resource: "albums", token: token, query: query)
        async let tracks: [Track] = fetchCollection(resource: "tracks", token: token, query: query)

        let results = try await (artists: artists, albums: albums, tracks: tracks)
        return CatalogResults(artists: results.artists, albums: results.albums, tracks: results.tracks)
    }

    private func fetchCollection<T: Decodable>(resource: String, token: String, query: String) async throws -> [T] {
        let trimmed = query.trimmingCharacters(in: .whitespacesAndNewlines)
        let queryItems: [URLQueryItem]? = trimmed.isEmpty ? nil : [
            URLQueryItem(name: "search", value: trimmed),
            URLQueryItem(name: "q", value: trimmed),
            URLQueryItem(name: "external", value: "true"),
        ]
        let response: PaginatedResponse<T> = try await client.send(
            "/api/v1/\(resource)/",
            token: token,
            queryItems: queryItems
        )
        return response.results
    }
}
